File: 백준/Gold/14938. 서강그라운드/서강그라운드.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;

const int INF = 1e9;
int n, m, r;
vector<int> t;
vector<vector<pii>> e;

int dji(int start){
    vector<int> dist(n+1, INF);
    dist[start] = 0;
    priority_queue<pii, vector<pii>, greater<pii>> pq; pq.push({0, start});
    while(pq.size()){
        auto [d, cur] = pq.top(); pq.pop();
        if(d > m) break;
        if(dist[cur] < d) continue;
        for(auto [nxt, nd] : e[cur]){
            int cost = d+nd;
            if(dist[nxt] > cost && m >= cost){
                dist[nxt] = cost;
                pq.push({cost, nxt});
            }
        }
    }
    int ret = 0;
    for(int i = 1; i <= n; i++) if(dist[i] <= m) ret += t[i-1];
    return ret;
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> n >> m >> r; e.resize(n+1); t.resize(n);
    for(int i = 0; i < n; i++) cin >> t[i];
    for(int i = 0; i < r; i++){
        int u, v, w; cin >> u >> v >> w;
        e[u].push_back({v, w});
        e[v].push_back({u, w});
    }
    int ans = 0;
    for(int i = 1; i <= n; i++) ans = max(ans, dji(i));
    cout << ans;
    return 0;
}

File: 백준/Gold/14938. 서강그라운드/서강그라운드.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

import heapq
from math import inf

def main():
    n, m, r = map(int, input().split())
    e = [[] for _ in range(n+1)]
    t = list(map(int, input().split()))
    for _ in range(r):
        u, v, w = map(int, input().split())
        e[u].append((v, w))
        e[v].append((u, w))
    def dji(start):
        dist = [inf]*(n+1)
        dist[start] = 0
        hq = [(0, start)]
        while hq:
            d, cur = heapq.heappop(hq)
            if d > m: break
            if dist[cur] < d: continue
            for nxt, nd in e[cur]:
                cost = d+nd
                if dist[nxt] > cost and m >= cost:
                    dist[nxt] = cost
                    heapq.heappush(hq, (cost, nxt))
        return sum(t[i-1] for i in range(1, n+1) if dist[i] <= m)
    ans = 0
    for i in range(1, n+1): ans = max(ans, dji(i))
    print(ans)
    return  
if __name__ == '__main__':
    main()